### PR TITLE
Static data parsing & performance optimization / measurement

### PR DESCRIPTION
**Pull request by Johannes-Schneider**

### CODE REVIEW

File: src/main/java/de/hpi/bpt/argos/util/performance/NullWatchTask.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package de.hpi.bpt.argos.util.performance;
+
+/**
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/util/performance/NullWatchTask.java]: MAJOR First sentence should end with a period. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AJavadoc_Style_BP2016)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/subParser/EntityTypeParser.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void endElement(String element) {
+		if (element.equalsIgnoreCase(NAME_ELEMENT) && !parentParser.latestOpenedElement(1).equalsIgnoreCase(ATTRIBUTE_ELEMENT)) {
+			if (!entityTypes.nameInUse(artifact.getName())) {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/subParser/EntityTypeParser.java]: MAJOR Merge this if statement with the enclosing one. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1066)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void startElement(String elementName) {
+		switch (elementName) {
+			case ENTITY_TYPES_ROOT_ELEMENT:
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java:194]: MAJOR Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+				return;
+
+			case ENTITY_INSTANCES_ROOT_ELEMENT:
+				activeSubParser = new EntityInstancesParser(this, entityTypes);
+
+				if (activeWatchTask != null) {
+					activeWatchTask.finish();
+				}
+
+				activeWatchTask = WatchImpl.start("parsing entity instances", () -> {});
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java]: MAJOR Either remove or fill this block of code. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00108)
MINOR '{' is not followed by whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)
MINOR '}' is not preceded with whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+		}
+
+		activeWatchTask.run(() -> activeSubParser.elementValue(latestOpenedElement(0), elementData));
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void endElement(String elementName) {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java]: MAJOR The Cyclomatic Complexity of this method "endElement" is 11 which is greater than 10 authorized. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+				if (activeSubParser instanceof EntityTypesParser) {
+					activeSubParser = null;
+				}
+				if (activeWatchTask != null) {
+					activeWatchTask.finish();
+				}
+				activeWatchTask = null;
+				break;
+
+			case ENTITY_INSTANCES_ROOT_ELEMENT:
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java]: MAJOR Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+	protected void startElement(String elementName) {
+		switch (elementName) {
+			case ENTITY_TYPES_ROOT_ELEMENT:
+				activeSubParser = new EntityTypesParser(this, entityTypes);
+
+				if (activeWatchTask != null) {
+					activeWatchTask.finish();
+				}
+
+				activeWatchTask = WatchImpl.start("parsing entity types", () -> {});
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java]: MAJOR Either remove or fill this block of code. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00108)
MINOR '{' is not followed by whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)
MINOR '}' is not preceded with whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void endElement(String elementName) {
+		if (activeSubParser == null) {
+			return;
+		}
+
+		switch (elementName) {
+			case ENTITY_TYPES_ROOT_ELEMENT:
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java]: MAJOR Reduce this switch case number of lines from 8 to at most 5, for example by extracting code into methods. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+				activeSubParser = new EntityTypesParser(this, entityTypes);
+
+				if (activeWatchTask != null) {
+					activeWatchTask.finish();
+				}
+
+				activeWatchTask = WatchImpl.start("parsing entity types", () -> {});
+				return;
+
+			case ENTITY_INSTANCES_ROOT_ELEMENT:
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java]: MAJOR Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/ArtifactParserImpl.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
+package de.hpi.bpt.argos.parsing.staticData;
+
+import de.hpi.bpt.argos.parsing.XMLFileParser;
+import de.hpi.bpt.argos.parsing.XMLSubParserImpl;
+
+/**
+ * {@inheritDoc}
+ * This is an abstract base class.
+ */
+public abstract class ArtifactParserImpl<ArtifactType> extends XMLSubParserImpl implements ArtifactParser<ArtifactType> {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/ArtifactParserImpl.java:10]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/ArtifactParserImpl.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+	protected ArtifactType parentArtifact;
+	protected ArtifactType artifact;
+	protected EntityTypeList entityTypes;
+
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/ArtifactParserImpl.java:15]: MAJOR Line has trailing spaces. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3ARegexp_Singleline_BP2016)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/ArtifactParser.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package de.hpi.bpt.argos.parsing.staticData;
+
+import de.hpi.bpt.argos.parsing.XMLSubParser;
+
+/**
+ * This interface offers methods to parse artifacts from a XML file.
+ */
+public interface ArtifactParser<ArtifactType> extends XMLSubParser {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/ArtifactParser.java:8]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java [hunk 13]
```diff
@@ -144,9 +146,51 @@ public boolean deleteArtifacts(PersistenceArtifact... artifacts) {
...
+																							 Class<ResultImplType> resultTypeClass,
+																							 Long... ids) {
+
+		ObjectWrapper<List<ResultType>> result = new ObjectWrapper<>();
+		WatchImpl.measure("get artifacts by id",
+				() -> result.set(executeGetArtifactsById(session, resultTypeClass, ids)));
+		return result.get();
+	}
+
+	/**
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java]: MAJOR First sentence should end with a period. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AJavadoc_Style_BP2016)

File: src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java [hunk 14]
```diff
@@ -144,9 +146,51 @@ public boolean deleteArtifacts(PersistenceArtifact... artifacts) {
 	 * {@inheritDoc}
 	 */
 	@Override
-	public <R, Q> R getArtifacts(Session session, Query<Q> query, Transaction transaction, Callable<R> getValue, R defaultValue) {
+	public <ResultType, QueryType> ResultType getArtifacts(Session session,
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java:149]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java [hunk 15]
```diff
@@ -144,9 +146,51 @@ public boolean deleteArtifacts(PersistenceArtifact... artifacts) {
...
+	 * @param session - the database session, which must be open
+	 * @param query - the query to execute and to retrieve the results from
+	 * @param transaction - the current transaction
+	 * @param getValue - the function to get the results from the query
+	 * @param defaultValue - a fall back default value in case anything went wrong or no entities were found
+	 * @param <ResultType> - the result type
+	 * @param <QueryType> - the query type
+	 * @return - an object of the result value type
+	 */
+	private <ResultType, QueryType> ResultType executeGetArtifacts(Session session,
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java:187]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java [hunk 16]
```diff
@@ -165,10 +209,15 @@ public boolean deleteArtifacts(PersistenceArtifact... artifacts) {
...
+	 * @param <ResultType> - the result type
+	 * @param <ResultImplType> - the implementation type of the result type
+	 * @param session - the database session, which must be open
+	 * @param resultTypeClass - the class of the result type
+	 * @param ids - the ids to fetch from the database
+	 * @return - a list of all entities, which were contained in the ids
 	 */
-	@Override
-	public <ResultType, ResultImplType extends ResultType> List<ResultType> getArtifactsById(Session session,
+	private <ResultType, ResultImplType extends ResultType> List<ResultType> executeGetArtifactsById(Session session,
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java:220]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java [hunk 17]
```diff
@@ -145,8 +146,19 @@ public String getParameter(String parameterName, boolean includePrefix) {
...
+		WatchImpl.measure(String.format("execute request: [%1$s] '%2$s'", request.requestMethod(), request.uri()),
+				() -> executeRequest(request,response, route));
+
+		logSendingResponse(logger, request, response.status(), response.body(), System.currentTimeMillis() - startMs);
+		WatchImpl.printResult(logger);
+
+		return response.body();
+	}
+
+	private void executeRequest(Request request, Response response, Route route) {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java]: MAJOR Missing a Javadoc comment. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AJavadocMethod_BP2016)

File: src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java [hunk 18]
```diff
@@ -145,8 +146,19 @@ public String getParameter(String parameterName, boolean includePrefix) {
 	 */
 	@Override
 	public String executeRequest(Logger logger, Request request, Response response, Route route) {
+		long startMs = System.currentTimeMillis();
 		logReceivedRequest(logger, request);
 
+		WatchImpl.measure(String.format("execute request: [%1$s] '%2$s'", request.requestMethod(), request.uri()),
+				() -> executeRequest(request,response, route));
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java]: MINOR ',' is not followed by whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_After_BP2016)

File: src/main/java/de/hpi/bpt/argos/parsing/util/ArtifactBatchImpl.java [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public long getTotalArtifactsStored() {
+		return totalArtifactsStored;
+	}
+
+	/**
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/util/ArtifactBatchImpl.java]: MAJOR First sentence should end with a period. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AJavadoc_Style_BP2016)

File: src/main/java/de/hpi/bpt/argos/parsing/util/ArtifactBatchImpl.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+		artifactsToStore = new ArrayList<>();
+		totalArtifactsStored = 0;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void setup() {
+		storageTask = WatchImpl.start("storing artifacts", () -> {});
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/util/ArtifactBatchImpl.java]: MAJOR Either remove or fill this block of code. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00108)
MINOR '{' is not followed by whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)
MINOR '}' is not preceded with whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)

File: src/main/java/de/hpi/bpt/argos/util/ObjectWrapper.java [hunk 21]
```diff
@@ -0,0 +1,26 @@
+package de.hpi.bpt.argos.util;
+
+/**
+ * This object is used to wrap another object.
+ * @param <ObjectType> - the type of the object to wrap
+ */
+public class ObjectWrapper<ObjectType> {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/util/ObjectWrapper.java:7]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/core/ArgosImpl.java [hunk 22]
```diff
@@ -49,8 +62,12 @@ public void start() {
 			return;
 		}
 
-		// TODO: parse static data
-		EventTypeParserImpl.getInstance().loadEventTypes();
+		WatchImpl.start("starting argos backend", () -> {});
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/core/ArgosImpl.java]: MAJOR Either remove or fill this block of code. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00108)
MINOR '{' is not followed by whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)
MINOR '}' is not preceded with whitespace. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3AWhitespace_Around_BP2016)

File: src/main/java/de/hpi/bpt/argos/util/performance/WatchImpl.java [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+		stop();
+		getWatch().logResult(logger);
+		watches.remove(Thread.currentThread().getId());
+	}
+
+	/**
+	 * This constructor initializes all members with their default values.
+	 */
+	private WatchImpl() {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/util/performance/WatchImpl.java]: MINOR Move this constructor to comply with Java Code Conventions. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1213)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 34 issues
* ![MAJOR][MAJOR] 15 major
* ![MINOR][MINOR] 19 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+	 */
+	@Override
+	protected void endElement(String elementName) {
+		if (activeSubParser == null) {
+			return;
+		}
+
+		switch (elementName) {
+			case ENTITY_TYPES_ROOT_ELEMENT:
+			case ENTITY_INSTANCES_ROOT_ELEMENT:
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java:250]: MAJOR Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

File: src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java [hunk 25]
```diff
@@ -0,0 +1,266 @@
...
+				activeSubParser = new EntityTypesParser(this, entityTypes);
+
+				if (activeWatchTask != null) {
+					activeWatchTask.finish();
+				}
+
+				activeWatchTask = WatchImpl.start("parsing entity types", () -> { /* empty, since only to group sub tasks */ });
+				return;
+
+			case ENTITY_INSTANCES_ROOT_ELEMENT:
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/parsing/staticData/StaticDataParserImpl.java:204]: MAJOR Reduce this switch case number of lines from 6 to at most 5, for example by extracting code into methods. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1151)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 16 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 10 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by hpibptlab** [comment]
SonarQube analysis reported 13 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 9 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Review by tommartensen — APPROVED** [review]

**Pull request merged**